### PR TITLE
[CRB-202] Add RegisterDealOrder command

### PR DESCRIPTION
**Pull request by nathanwhit**
This PR implements the RegisterDealOrder command, which creates an simpler entry point for recording loans.

This is based on the changes from #8 and #12, so this is blocked on both of those PRs being merged.

### CODE REVIEW

**Review by atodorov — CHANGES_REQUESTED** [review]
@nathanwhit one of the negative test cases I'm working on fails with a slightly mismatching error message
```
  X InvalidExpiration_Neg [47ms]
  Error Message:
     Expected string length 60 but was 53. Strings differ at index 53.
  Expected: "Transaction status: INVALID;  Message: Invalid number format"
  But was:  "Transaction status: INVALID;  Message: Invalid number"
```

All of the others use "Invalid number format". I didn't see this in the diff but maybe you know where it is coming from. I advocate for consistency. More tests to come soon. See https://github.com/gluwa/Creditcoin-Tests/pull/25

**Comment by nathanwhit** [comment]
> 
> 
> @nathanwhit one of the negative test cases I'm working on fails with a slightly mismatching error message
> 
> ```
>   X InvalidExpiration_Neg [47ms]
>   Error Message:
>      Expected string length 60 but was 53. Strings differ at index 53.
>   Expected: "Transaction status: INVALID;  Message: Invalid number format"
>   But was:  "Transaction status: INVALID;  Message: Invalid number"
> ```
> 
> All of the others use "Invalid number format". I didn't see this in the diff but maybe you know where it is coming from. I advocate for consistency. More tests to come soon. See [gluwa/Creditcoin-Tests#25](https://github.com/gluwa/Creditcoin-Tests/pull/25)

So the `Invalid number` message is a quirk from the C++ transaction processor. It also appears in the [addOffer tests](https://github.com/gluwa/Creditcoin-Tests/blob/c57fd7265fa0117fe7e6e6bfd86067126facff0f/Creditcoin.NUnitTests/AddOfferTests.cs#L126-L134). I agree with making the message consistent, I'll go ahead and make that change.

File: src/handler.rs [hunk 1]
```diff
@@ -1245,6 +1353,242 @@ impl CCTransaction for CloseDealOrder {
...
+        let transfer_id_key = string!(
+            &ask_address.blockchain,
+            &self.blockchain_tx_id,
+            &ask_address.network
+        );
+        let transfer_id = Address::with_prefix_key(TRANSFER, &transfer_id_key);
+
+        if let Some(transfer) = try_get_state_data(tx_ctx, &transfer_id)? {
+            bail_transaction!(
+                "The transaction has already been registered",
```
Comment (atodorov) [inline src/handler.rs]: @nathanwhit according to the ticket description the message should be "The transfer has already been registered". Let me know if you'd like to adjust it or if I should do it on my end. 

```
  Failed TransferAlreadyRegistered_Neg [5 m 34 s]
  Error Message:
     Expected string length 79 but was 82. Strings differ at index 48.
  Expected: "...NVALID;  Message: The transfer has already been registered"
  But was:  "...NVALID;  Message: The transaction has already been registered"
  -----------------------------------------^
```

File: src/handler.rs [hunk 2]
```diff
@@ -1245,6 +1353,242 @@ impl CCTransaction for CloseDealOrder {
...
+            &self.blockchain_tx_id,
+            &ask_address.network,
+        ]
+        .join(" ");
+        ctx.verify(&gateway_command)?;
+
+        let deal_order = protos::DealOrder {
+            blockchain: ask_address.blockchain,
+            src_address: ask_address.value,
+            dst_address: bid_address.value,
```
Comment (atodorov) [inline src/handler.rs]: @nathanwhit I think these should be `ask_address_id` and `bid_address_id`, see the details at
https://github.com/gluwa/Creditcoin-Tests/pull/25#discussion_r728375218

**Review by atodorov — CHANGES_REQUESTED** [review]

File: src/handler.rs (hunk 2 above)
Comment (nathanwhit) [reply src/handler.rs]: Yup you're right, fixed now.

File: src/handler.rs (hunk 1 above)
Comment (nathanwhit) [reply src/handler.rs]: Fixed in latest commit, thanks for pointing it out.

**Review by atodorov — APPROVED** [review]
After the latest changes all tests from https://github.com/gluwa/Creditcoin-Tests/pull/25 report PASS.

**Comment by atodorov** [comment]
@nathanwhit , @CAGS295 ping. Can we merge this PR? I don't see a reason for it staying open. Previous comment indicates tests have passed (I had tested against a custom built docker image).

**Comment by atodorov** [comment]
@nathanwhit #8 and #20 (replacement for #12) have been now merged. Can you please rebase this PR so I can rebuild the docker image & retest it.

**Comment by nathanwhit** [comment]
@atodorov Rebased

**Review by seandeneen-gluwa — APPROVED** [review]

**Review by CAGS295 — APPROVED** [review]

**Comment by atodorov** [comment]
After additional investigation of change deltas we agree that this looks good. Merging.

**Pull request merged**